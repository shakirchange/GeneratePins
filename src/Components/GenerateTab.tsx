import React from 'react';
import * as actionTypes from '../Action/action';
import { connect } from 'react-redux';
import Classes from '../Components/generatepin.module.css';
export interface PropsValues {
    generateRamdonNumbers: number[];
    setGeneratedPins: any;
    savedPinsArray: any;
    namesArray: any;
}
export class GenerateTab extends React.Component<PropsValues>{
    constructor(props: any) {
        super(props);
        this.state = {
            generateRamdonNumbers: []
        }
    }
    checkNumberSequence3digits = (num: number) => {
        let numStr = num + "";
        let first3String: string = numStr.substring(0, 3);
        let second3String: string = numStr.substring(1, 4);
        let revNum: string = this.reverseString(numStr);
        let first3StringREV = revNum.substring(0, 3);
        let second3StringREV = revNum.substring(1, 4);
        if (this.check3digitsSequence(first3String) || this.check3digitsSequence(second3String)
            || this.check3digitsSequence(first3StringREV) || this.check3digitsSequence(second3StringREV)) {
            return true;
        }
        return false;
    }

    check3digitsSequence = (num: string) => {
        let firstDigit = num.charAt(0);
        let secondDigit = num.charAt(1);
        let thirdDigit = num.charAt(2);
        let firstNum = parseInt(firstDigit + "");
        let secondNum = parseInt(secondDigit + "");
        let thirdNum = parseInt(thirdDigit + "");
        if (firstNum + 1 === secondNum && firstNum + 2 === thirdNum) {
            return true;
        }
        return false;
    }

    reverseString = (str: string) => {
        if (!str || str.length < 2 ||
            typeof str !== 'string') {
            return 'Not valid';
        }
        const reletray = [];
        const length = str.length - 1;
        for (let i = length; i >= 0; i--) {
            reletray.push(str[i]);
        }
        return reletray.join('');
    }
    check2NumbersDuplicate = (num: number) => {
        let strNum: string = num + "";
        let flag: boolean = false;
        for (let i = 0; i < strNum.length; i++) {
            if (i !== strNum.length - 1) {
                if (strNum[i] === strNum[i + 1]) {
                    flag = true;
                    break;
                }
            }
        }
        return flag;
    }
    generate4digitPIN = () => {
        let i: number = 0;
        let arrayLen: number[] = [];
        while (i <= 4) {
            let generatedPassword = Math.floor(1000 + Math.random() * 9000);
            let genPassword: string = generatedPassword + "";
            if (genPassword.charAt(0) === "0") continue;
            let num = parseInt(genPassword);
            if (!this.check2NumbersDuplicate(num) && !this.checkNumberSequence3digits(num) && !arrayLen.includes(num)) {
                arrayLen[i] = num;
                i++;
            }
        }
        this.props.setGeneratedPins(arrayLen, "generateRamdonNumbers");
    }
    componentDidMount() {
        this.generate4digitPIN();
    }
    save4digitPIN = () => {
        let generatedPins = this.props.generateRamdonNumbers;
        let savedPinsArray = this.props.savedPinsArray;
        if (savedPinsArray.data && savedPinsArray.data.includes(generatedPins)) {
            alert("Pins are already saved");
        } else {
            savedPinsArray.data.push(generatedPins);
            this.props.setGeneratedPins(savedPinsArray, "savedPinsArray");
            alert("Pins Saved Successfully");
        }
    }
    render() {
        const array = this.props.generateRamdonNumbers;
        const generatedPins = array.map((data, index) => {
            return (
                <li key={index}>
                    <input type="text" defaultValue={data} readOnly />
                </li>
            )
        })
        return (
            <React.Fragment>
                <ul className={Classes.textAlignment}>
                    {generatedPins}
                </ul>
                <div className={Classes.textAlignment}>
                    <button id="generate" className={Classes.success1} onClick={this.generate4digitPIN}>Generate</button>
                    <button id="save" className={Classes.success} onClick={this.save4digitPIN} >Save</button>
                </div>
            </React.Fragment>
        )
    }
}
const mapStateToProps = (state: any) => {
    return {
        generateRamdonNumbers: state.GeneratePinsReducer.generateRamdonNumbers,
        savedPinsArray: state.GeneratePinsReducer.savedPinsArray,
        namesArray: state.GeneratePinsReducer.namesArray
    }
}
const mapDispatchToProps = (dispatch: any) => {
    return {
        setGeneratedPins: (fieldValue: any, fieldName: 'string') => {
            dispatch({ type: actionTypes.GENERATE_ON_CLICK_HANDLER, fieldValue, fieldName });
        }
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(GenerateTab)