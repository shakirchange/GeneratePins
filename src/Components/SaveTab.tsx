import React from 'react';
import * as actionTypes from '../Action/action';
import { connect } from 'react-redux';
import Classes from '../Components/generatepin.module.css';
import { SavePinsInterface } from '../Common/GeneratePinsInterface';

export class SaveTab extends React.Component<SavePinsInterface>{
    constructor(props:any){
        super(props);
        this.state= {
            name: ''
        }
    }
    //Deleting seleted row function
    deleteRow = (e: any, i: number) => {
        let savedPinsArr = this.props.savedPinsArray;
        let namesArr = this.props.namesArray;
        if (savedPinsArr.data && savedPinsArr.data.length > 0) {
            savedPinsArr.data.splice(i, 1);
        }
        if (namesArr && namesArr.length > 0) {
            namesArr.splice(i, 1);
        }
        this.setState({ name: '' });
        this.props.deletePins(savedPinsArr, "savedPinsArray");
        this.props.deletePins(namesArr, "namesArray");
        alert("Selected Row deleted successfully");
    }
    //Setting the name function
    setName = (e: any, i: number) => {
        let namesArr = this.props.namesArray;
        namesArr[i] = e.target.value;
        this.props.setGeneratedPins(namesArr, "namesArray");
    }
    //Building saved pins html
    buildSaveTabsUI = (pinsObj: any, objArr: any, i: number) => {
        let namesArr = this.props.namesArray;
        let name: string = "";
        if (namesArr && namesArr.length > 0) {
            name = namesArr[i];
        }
        pinsObj.push(<li key={"name_" + i} className={Classes.li}><input className={Classes.nameInputtext} type="name" id={"name" + i} defaultValue={name || ""} onChange={(e) => this.setName(e, i)} /></li>);
        for (let j = 0; j < objArr.length; j++) {
            let obj = objArr[j];
            if (obj) {
                pinsObj.push(<li key={"input_" + i + "_" + j} className={Classes.li}><input defaultValue={obj} readOnly /></li>);
            }
        }
        pinsObj.push(<li key={"delete_" + i} className={Classes.li}><button className={Classes.delete} onClick={(e) => this.deleteRow(e, i)}>DELETE</button></li>);
        pinsObj.push(<br key={"newline_" + i} />);
        return pinsObj;
    }
    render() {
        let savedPinsArray = this.props.savedPinsArray;
        let pinsObj: string[] = [];
        if (savedPinsArray && savedPinsArray.data && savedPinsArray.data.length > 0) {
            for (let i = 0; i < savedPinsArray.data.length; i++) {
                let objArr = savedPinsArray.data[i];
                if (objArr && objArr.includes(',')) {
                    objArr = objArr.split(",");
                    if (objArr && objArr.length > 0) {
                        pinsObj = this.buildSaveTabsUI(pinsObj, objArr, i);
                    }
                } else {
                    pinsObj = this.buildSaveTabsUI(pinsObj, objArr, i);
                }
            }
        }
        return (
            <React.Fragment>
                <div className={Classes.textAlignment}>
                    <ul className={Classes.ul}>
                        {pinsObj.length > 0 ? pinsObj : <li>No Saved Pins found</li>}
                    </ul>
                </div>
            </React.Fragment>
        )
    }
}
const mapStateToProps = (state: any) => {
    return {
        generateRamdonNumbers: state.GeneratePinsReducer.generateRamdonNumbers,
        savedPinsArray: state.GeneratePinsReducer.savedPinsArray,
        pinsObj: state.GeneratePinsReducer.pinsObj,
        namesArray: state.GeneratePinsReducer.namesArray
    }
}
const mapDispatchToProps = (dispatch: any) => {
    return {
        setGeneratedPins: (fieldValue: any, fieldName: 'string') => {
            dispatch({ type: actionTypes.GENERATE_ON_CLICK_HANDLER, fieldValue, fieldName });
        },
        deletePins: (fieldValue: any, fieldName: 'string') => {
            dispatch({ type: actionTypes.DELETE_ON_CLICK_HANDLER, fieldValue, fieldName });
        }
    }
}
export default connect(mapStateToProps, mapDispatchToProps)(SaveTab)