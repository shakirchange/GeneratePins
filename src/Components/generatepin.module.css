
.textAlignment {
    text-align: center;
}
.row {
    position: relative;
    top: 100px;
}
.generateButton{
    margin: 5px;
    padding: 10px;
    border-radius: 5px;
    color: hsl(208, 81%, 63%);
    width: 160px;
    border: solid 1px #e2e2e2;
    margin-left: 45px;
    border: solid 1px hsl(208, 81%, 63%);
    cursor: pointer;
}
.nameInputtext{
    width: 150px;;
}
.saveButton{
    margin: 5px;
    padding: 10px;
    border-radius: 5px;
    background: hsl(208, 81%, 63%);
    color: white;
    width: 160px;
    border: solid 1px #e2e2e2;
    cursor: pointer;
}
.delete{
    margin: 5px;
    padding: 10px;
    border-radius: 5px;
    background: red;
    color: white;
    width: 100px;
    border: solid 1px #e2e2e2;
    cursor: pointer;
}
ul, .ul{
    list-style-type: none; 
}
ul > li, .ul> .li{
    display: inline-flex !important;
}
input{
    margin: 5px;
    padding: 5px;
    height: 20px;
    width: 50px;
    border: solid 1px black;
    
}