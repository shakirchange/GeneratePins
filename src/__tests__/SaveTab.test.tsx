import React from 'react';
import { shallow } from 'enzyme';
import { SaveTab } from '../Components/SaveTab';
import EnzymeAdapter from 'enzyme-adapter-react-16';
import { configure } from 'enzyme';
configure({ adapter: new EnzymeAdapter() });
let PropsValues= {
    generateRamdonNumbers: [],
    savedPinsArray: [],
    namesArray: [],
    deletePins: [],
}
describe('render generate tan', () => {
    let wrapper: any;
    let setGeneratedPins = jest.fn();
    wrapper = shallow(<SaveTab
        setGeneratedPins={setGeneratedPins}
        {...PropsValues} />)
    test('Testing Snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })
    test('Testing Length', () => {
        expect(wrapper.length).not.toBe(null);
    })
    test('Testing Divs Length', () => {
        expect(wrapper.find('div').length).toEqual(1);
    })
    test('Testing ul Elements Length', () => {
        expect(wrapper.find('ul').length).toEqual(1);
    })
})