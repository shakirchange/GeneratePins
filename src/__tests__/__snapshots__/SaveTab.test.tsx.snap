// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render generate tan Testing Snapshot 1`] = `ShallowWrapper {}`;
